### ISSUE TITLE
Rule Change: support `Array.fromAsync` in `array-callback-return`

### ISSUE THREAD

**Issue by fasttime** [eslint/eslint#20438]
### What rule do you want to change?

array-callback-return

### What change do you want to make?

Generate more warnings

### How do you think the change should be implemented?

A new option

### Example code

```js
/* eslint array-callback-return: "error" */

const results = Array.fromAsync(
    values,
    value => { not_a_property: value }
);
```

[**Playground link**](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IGFycmF5LWNhbGxiYWNrLXJldHVybjogXCJlcnJvclwiICovXG5cbmNvbnN0IHJlc3VsdHMgPSBBcnJheS5mcm9tQXN5bmMoXG4gICAgdmFsdWVzLFxuICAgIHZhbHVlID0+IHsgbm90X2FfcHJvcGVydHk6IHZhbHVlIH1cbik7Iiwib3B0aW9ucyI6eyJydWxlcyI6e30sImxhbmd1YWdlT3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e319fX19)

### What does the rule currently do for this code?

No problem is currently reported.

### What will the rule do after it's changed?

It will report a problem when the callback passed to `Array.fromAsync` does not return a value, consistent with the behavior for `Array.from`.

In addition, because `Array.fromAsync` commonly uses async callbacks, it may be useful for the rule to also flag async functions that do not return a value. At the moment, async callbacks are ignored, but for `Array.fromAsync` this is likely the more common case.

```js
// This is an error:
const a = Array.from(foo, bar => { /* no return */ });

// This should be also an error:
const b = Array.fromAsync(foo, bar => { /* no return */ });

// This is not an error:
const c = Array.from(foo, async bar => { /* no return */ });

// But maybe this should be an error?:
const d = Array.fromAsyc(foo, async bar => { /* no return */ });
```

[**Playground link**](https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgYXJyYXktY2FsbGJhY2stcmV0dXJuOiBcImVycm9yXCIqL1xuXG4vLyBUaGlzIGlzIGFuIGVycm9yOlxuY29uc3QgYSA9IEFycmF5LmZyb20oZm9vLCBiYXIgPT4geyAvKiBubyByZXR1cm4gKi8gfSk7XG5cbi8vIFRoaXMgc2hvdWxkIGJlIGFsc28gYW4gZXJyb3I6XG5jb25zdCBiID0gQXJyYXkuZnJvbUFzeW5jKGZvbywgYmFyID0+IHsgLyogbm8gcmV0dXJuICovIH0pO1xuXG4vLyBUaGlzIGlzIG5vdCBhbiBlcnJvcjpcbmNvbnN0IGMgPSBBcnJheS5mcm9tKGZvbywgYXN5bmMgYmFyID0+IHsgLyogbm8gcmV0dXJuICovIH0pO1xuXG4vLyBCdXQgbWF5YmUgdGhpcyBzaG91bGQgYmUgYW4gZXJyb3I/OlxuY29uc3QgZCA9IEFycmF5LmZyb21Bc3ljKGZvbywgYXN5bmMgYmFyID0+IHsgLyogbm8gcmV0dXJuICovIH0pOyIsIm9wdGlvbnMiOnsicnVsZXMiOnt9LCJsYW5ndWFnZU9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnt9fX19fQ==)

### Participation

- [x] I am willing to submit a pull request to implement this change.

### Additional comments

[`Array.fromAsync`](https://github.com/tc39/proposal-array-from-async#readme) has reached TC39 stage 4, so we should support it in core rules. See also #20437.

**Comment by DMartens** [comment]
This makes sense as we already support `Array.from` and I agree that also `async` callbacks without return value should be reported.

**Comment by mdjermanovic** [comment]
> It will report a problem when the callback passed to `Array.fromAsync` does not return a value, consistent with the behavior for `Array.from`.

Makes sense to me.

> In addition, because `Array.fromAsync` commonly uses async callbacks, it may be useful for the rule to also flag async functions that do not return a value. At the moment, async callbacks are ignored, but for `Array.fromAsync` this is likely the more common case.

This also makes sense to me. Per https://github.com/tc39/proposal-array-from-async?tab=readme-ov-file#mapping-function, each result of the mapping callback is awaited, so it doesn't seem like there could be a valid use case for not returning a value, unlike other array methods this rule checks (e.g, `Promise.all(list.map(async x => { await doSomething(x) })` is valid, as explained in https://github.com/eslint/eslint/issues/7101).

Marking as accepted.

**Comment by fasttime** [comment]
I will work on a PR.
